Ensure null-termination of filenames copied by strncpy

archigup opened the pull request:
Prior use of strncpy passed the buffer length for n; when the string to copy is
length n or longer, this results in the buffer not containing a null byte. This
is problematic as the buffers are assumed to contain a null byte elsewhere.

Reported here: https://forums.freertos.org/t/freertos-fat-compiler-warnings/14843

htibosch reviewed (commented):
Hello @archigup, thank you for this PR!

The reason that you see this existing code:
~~~c
    char pcFileName[ ffconfigMAX_FILENAME ];
~~~
is that:

- all filename access is supposed to be done by STRNCPY(), which ensures a trailing zero.
- it is assumed that the zero termination byte is included in `ffconfigMAX_FILENAME`.

I often have `ffconfigMAX_FILENAME` defined as `128+1` or `256+1`.

I am not a fan of the `= { 0 }` initialisation for two reasons:
- The compiler will do lots of useless `memset`'s, thus losing CPU time
- The implementation of these `memset`'s is hidden, and on some platforms can lead to a crash. E.g. it is possible that the compiler uses wide FPU registers for this, while they're not saved in the task context. Another problem occurs when the compiler makes wrong assumptions about the alignment of the variable.

It doesn't look as beautiful, but in stead I would like to see:
~~~
    char pcFileName[ ffconfigMAX_FILENAME + 1 ];
    pcFileName[ sizeof( pcFileName ) - 1 ] = 0;
~~~

which is more efficient, and it has a very clear implementation.

And this one:

> `FF_DirEnt_t xDirEntry = { 0 };`

The same comment here. I would rather call `memset()` explicitly.
But why would you clear it, are the fields not set by the code? If so, do you have an example for that?

archigup commented:
Thanks for the feedback. I have updated the PR.

I had previously assumed `ffconfigMAX_FILENAME`  did not include null byte; I have updated PR accordingly.

STRNCPY does not ensure a trailing zero. STRNCPY just calls strncpy. As noted in the GCC warning we are passing strncpy a length equal to the buffer size.

From the strncpy man page we can see:
```
The strncpy() function is similar, except that at most n bytes of src are copied.  Warning: If there is no null byte among the first n bytes of src, the string placed in dest will not be null-terminated.
```
and
```
If there is no terminating null byte in the first n bytes of src, strncpy() produces an unterminated string in dest.  If buf has length buflen, you can force termination using something like the following:

           if (buflen > 0) {
               strncpy(buf, str, buflen - 1);
               buf[buflen - 1]= '\0';
           }
```

As such, the current uses of strncpy will end up with no null termination if the copied string is length greater than (ffconfigMAX_FILENAME - 1).

The PR has been updated to just fix uses of strncpy. I have also updated it to use ` pcFileName[ sizeof( pcFileName ) - 1 ] = 0;` instead of `= { 0 }` to ensure null bytes.

htibosch commented on ff_dir.c:
```diff
@@ -1723,9 +1723,11 @@ FF_Error_t FF_PopulateLongDirent( FF_IOManager_t * pxIOManager,
                         #if ( ffconfigUNICODE_UTF16_SUPPORT != 0 )
-                            wcsncpy( pxDirEntry->pcWildCard, ++pcWildCard, ffconfigMAX_FILENAME );
+                            wcsncpy( pxDirEntry->pcWildCard, ++pcWildCard, ffconfigMAX_FILENAME - 1 );
+                            pxDirEntry->pcWildCard[ sizeof( pxDirEntry->pcWildCard ) - 1 ] = 0;
```
You wrote:

> `pxDirEntry->pcWildCard[ sizeof( pxDirEntry->pcWildCard ) - 1 ] = 0;`

Now that I see the new code, I realise that `sizeof( pcWildCard  )` isn't equal to the number of elements in an array of `wchar_t`.
I think that we should write:
~~~diff
-  pxDirEntry->pcWildCard[ sizeof( pxDirEntry->pcWildCard ) - 1 ] = 0;
+  pxDirEntry->pcWildCard[ ffconfigMAX_FILENAME ] = 0;
~~~
In ff_dir.c, `strncpy()` and `wcsncpy()` are called, depending on `ffconfigUNICODE_UTF16_SUPPORT`, whereas in ff_file.c, the macro `STRNCPY` is used.
What about having ff_dir.c use that macro as well?

PS. My proposal to use `sizeof` as an index is that it is never mistaken about the actual length of a string. Except when it is a string of 16-bit entries...

htibosch reviewed (changes requested):
I'm glad with the changes, very good.

I would like to propose one more change: use the macro `STRNCPY()` in  ff_dir.c, just as it is used in ff_file.c

archigup commented:
Updated PR to use the macro in ff_dir.c and to use the length to index instead of sizeof

htibosch reviewed (approved):
It looks all good to me, and I approve this PR.
Thank you [Archit](https://github.com/archigup).
